### ISSUE TITLE
Two Finger scrolling?

### ISSUE THREAD

**Issue by withinboredom** [danielquinn/Gentoo-Surface-Pro-3#14]
Has anyone gotten two-finger scrolling to work -- or maybe I'm just missing something obvious?

**Comment by danielquinn** [comment]
It works just fine for me.  Make sure you've done the following:

1. Applied the two multitouch patches in the `kernel` directory of this repo
2. Installed `dev-libs/libinput` if you're running GNOME (I don't know about other DEs)

If that doesn't work, try configuring your kernel with the `.config` file in the `kernel` directory of this repo.  At that point, you'd be running basically everything I'm using.

**Comment by withinboredom** [comment]
Done, done and done ;) You were my inspiration to finally make the switch ...

alas, I apparently have the second generation type cover (my 1st gen suffered under a flood of coffee caused by a 2 year old). So, I tried this:

``` diff
Section "InputClass"
        Identifier "Surface Pro 3 cover"
        MatchIsPointer "on"
        MatchDevicePath "/dev/input/event*"
        Driver "evdev"
        Option "vendor" "045e"
-        Option "product" "07dc"
+        Option "product" "07e2"
        Option "IgnoreAbsoluteAxes" "True"
EndSection
```

This still didn't enable multi-touch, but I have a feeling I'm on the right path. I'll report back here with anything I find...

**Comment by withinboredom** [comment]
If you remove your type cover and then reattach it, what does your `/var/log/Xorg.0.log` say?

mine:

```
[  1991.973] (II) config/udev: Adding input device Microsoft Surface Type Cover (/dev/input/mouse2)
[  1991.973] (**) Microsoft Surface Type Cover: Applying InputClass "system-keyboard"
[  1991.973] (II) No input driver specified, ignoring this device.
[  1991.973] (II) This device may have been added with another device file.
[  1991.986] (II) config/udev: Adding input device Microsoft Surface Type Cover (/dev/input/event10)
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "evdev pointer catchall"
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "evdev keyboard catchall"
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "libinput pointer catchall"
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "libinput keyboard catchall"
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "system-keyboard"
[  1991.986] (**) Microsoft Surface Type Cover: Applying InputClass "Surface Pro 3 cover"
```

**Comment by withinboredom** [comment]
After running evtest and exev, I can see the mousewheel events are coming from the hardware but not making it to X. Hmm.

After looking at the output from `/dev/input/mouse2` I can see that it is not getting mousewheel events from evdev. My suspicion is that gnome is using `/dev/input/mouse2` instead of the `dev/input/event*` ... if I'm reading the log right, it might be using both, but the mouse2 is overriding the evdev.

**Comment by danielquinn** [comment]
What's your `INPUT_DEVICES` value in `/etc/portage/make.conf`?  This is mine:

    INPUT_DEVICES="libinput keyboard mouse"

...but I think even just `"libinput"` is enough.

**Comment by withinboredom** [comment]
I apparently read the output of evtest wrong, it's not outputting the mousewheel events at all.

**Comment by withinboredom** [comment]
ahhh ... I had syntaptics and evdev, but no libinput :man_facepalming: for some reason ... rebuilding the world. I'll report back once complete.

**Comment by danielquinn** [comment]
Just a note that you shouldn't need to rebuild all of `@world` since you're only talking about X modules, you can do `emerge @x11-modules`.  It's way faster :-)

**Comment by withinboredom** [comment]
Quick question, do you have issues with putting the device to sleep and gnome? I get lots of flickering of the backlight and it never comes back.

**Comment by danielquinn** [comment]
OMG yes.  Sleeping has never worked well for me.  It'll lock, the screen will turn off and the fan will stop... for about a minute.  Then the screen comes back, with the fan, and the battery just drains until I realise what's happening.  I have no idea why, but to be honest, I've always had trouble with hibernation and Linux.

**Comment by withinboredom** [comment]
Soo ... still doesn't work. It might be that this version of the cover doesn't work well with linux (I'd throw a bet that this is on purpose?) and needs some further kernel modifications to get to work. I just pulled out my v1 typecover and it works flawlessly, other than over half the keyboard doesn't work due to coffee damage, so it has to be hardware related...

Good to know I'm not the only one with sleeping problems :p

**Comment by danielquinn** [comment]
Well I live in the UK, so I've been up for a while ;-)

As for the keyboard support, I guess it depends on what you're trying to use.  My Surface Pro 3 works just fine in Linux with the TypeCover 3 that I bought with it *and* with the TypeCover 4 that I bought to replace it 'cause it's *so* much better.

**Comment by withinboredom** [comment]
Haha, I'll just get a TypeCover 4. Do I see a `ins` key on there (yes please!)? What's that right beside the left-arrow key?

**Comment by danielquinn** [comment]
Yup.  That was one of the main reasons I bought it.  A proper insert key and a bigger, smoother touchpad.  They keys are nicely separated (better for touch typers like me) and it just *feels* nicer to use.  I highly recommend an upgrade :-)

**Comment by withinboredom** [comment]
Looks like full support is coming in 4.10 merge window ... :D

**Comment by danielquinn** [comment]
Dear gods I hope so.  Maintaining this repo has been a pain in the ass :-)

**Comment by withinboredom** [comment]
lol, This is in the 4.9.5 kernel:

```
CONFIG_TOUCHSCREEN_SUR40=m
CONFIG_TOUCHSCREEN_SURFACE3_SPI=m
```

**Comment by withinboredom** [comment]
FWIW: Try gentoo-sources-4.9.6 without patches. Works out of the box (except multi-touch mouse -- but may work with SP4 keyboard)
